[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110540

With use_orig_params=True, it is possible that some parameters with the same FlatParameter are in the optimizer while others parameters are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

fegin committed:
[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models

With use_orig_params=True, it is possible that some parameters in a FlatParameter are in the optimizer while others are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters in a FlatParameter are in the optimizer while others are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters in a FlatParameter are in the optimizer while others are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters with the same FlatParameter are in the optimizer while others parameters are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1609,7 +1624,19 @@ def _gather_all_orig_param_state(
+                f"{key} is not in the output state. "
+                "The FSDPParamInfo has the param keys "
+                f"{sorted(list(fsdp_param_info.param_indices.keys()))} while "
+                "the output_states doesn't have the param keys "
```
good catch!

rohan-varma reviewed (approved):
Similar question to @awgu - what happens when a parameter transitions state from frozen -> nonfrozen, or vice-versa during training?

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters with the same FlatParameter are in the optimizer while others parameters are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin commented:
The latest version supports the no_grad -> grad -> no_grad transition and changes the UT to verify the usage.

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters with the same FlatParameter are in the optimizer while others parameters are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Make the new optimizer allgather fusion work with fine-tuning models"


With use_orig_params=True, it is possible that some parameters with the same FlatParameter are in the optimizer while others parameters are frozen. This PR makes the allgather fusion logic support the case.

Differential Revision: [D49922028](https://our.internmc.facebook.com/intern/diff/D49922028/)

[ghstack-poisoned]
